template/{% if use_git and dev_platform == 'GitHub' %}.github{% endif %}/workflows/release.yml

```diff
@@ -1,53 +1,74 @@
-name: CD
+name: Publish to PyPI
 
 on:
-  push:
-    tags:
-    - 'v?[0-9]+.[0-9]+.[0-9]+'
-    - 'v?[0-9]+.[0-9]+.[0-9]+(a|b|rc|post|dev)[0-9]+'
-
-jobs:
-  prerequisites:
-    uses: ./.github/workflows/test.yml
-
   release:
-    needs: [prerequisites]
-    strategy:
-      matrix:
-        os: [ubuntu-latest]
-        python-version: ["3.12"]
-    runs-on: ${{ matrix.os }}
-    permissions:
-      # Write permissions are needed to create OIDC tokens.
-      id-token: write
-      # Write permissions are needed to make GitHub releases.
-      contents: write
+    types: [published]
 
+jobs:
+  # Setup build separate from publish for added security
+  # See https://github.com/pypa/gh-action-pypi-publish/issues/217#issuecomment-1965727093
+  build:
+    runs-on: ubuntu-latest
+    # This ensures that the publish action only runs in the main repository
+    # rather than forks
+    # Environment is encouraged for increased security
+    environment: build
```

Comment:
This is a new one for me, so I'm curious. Could you please share how it works?

Reply:
Oh! @agriyakhetarpal environments allow you to setup a trusted connection with pypi. This is the build environment, so we don't need to worry too much about security here in terms of PYPI. so anyone can run the build step in theory. but by having a publish environment, we can control who can trigger that step of the build.  (if that is what you're asking) . 

here is a bit more about this

https://docs.pypi.org/trusted-publishers/using-a-publisher/

Reply:
Oh, I'm aware of trusted publishing and environments, thanks! My question was more around the need for this `build` environment, as it seems to be redundant. Most release workflows can get by with just one `pypi` (or `publish`, or any other similar name) environment, omitting the need for an environment for the build step. Does including two environments, one for building and one for publishing, have security provisions? Moreover, I imagine that the users would also have questions around why they need (and we recommend) to set up two environments in their repository's settings, when it is only the publishing one that connects to PyPI's OIDC publishing functionalities.

Reply:
Oh gosh my apologies. i suspected I was telling you something you already knew. 🙈 

This is a great point. I don't know the answer to your question!! But i suspect this is ME introducing complexity. 

i just checked the [tutorials](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/) and there is a build step but not an additional environment. Shall we remove this environment as you are correct it's not required. we need the job but not an environment AFAIK .